jax/experimental/compilation_cache/gfile_cache.py

```diff
@@ -0,0 +1,45 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+
+from jax.experimental.compilation_cache.cache_interface import CacheInterface
+import tensorflow.io.gfile as gfile
+from absl import logging
+
+class GFileCache(CacheInterface):
+
+  def __init__(self, path: str):
+    """Sets up a cache at 'path'. Cached values may already be present."""
+    self._path = pathlib.Path(path)
+    gfile.makedirs(self._path)
+
+  def get(self, key: str):
+    """Returns None if 'key' isn't present."""
+    if not key:
+      raise ValueError("key cannot be empty")
+    path_to_key = str(self._path / key)
+    if gfile.exists(path_to_key):
+      with gfile.GFile(path_to_key, "rb") as file:
+        return file.read()
+    else:
+      return None
+
+  def put(self, key: str, value: bytes):
+    """Adds new cache entry."""
+    if not key:
+      raise ValueError("key cannot be empty")
+    path_to_new_file = str(self._path / key)
+    with gfile.GFile(path_to_new_file, "w") as f:
+      f.write(value)
```

Comment:
We need to atomically create the file here. I think GCS renames are atomic, so I think it would work to write a temporary version of the file, maybe something like `tmp_path = str(self._path / "_" + key)` and then move that to `path_to_new_file`.

Reply:
GCS renames are NOT atomic.

GCS does a copy and then delete.

![image](https://user-images.githubusercontent.com/14262417/171503424-b581ad07-ac1f-4b7b-a618-476e185bcf77.png)

Reply:
Oh wow. If they're immutable, hopefully this means the copy is effectively atomic though? And maybe the write itself? Like it seems like it should be impossible to read a half-written file.

Reply:
micban@ told us to not depend on rename in GCS at all. I'll tag you in that thread.